packages/shopping/src/__tests__/acceptance/user.controller.acceptance.ts

```diff
@@ -100,106 +96,162 @@ describe('UserController', () => {
   });
 
   it('throws error for POST /users with a string', async () => {
-    await client
+    const res = await client
       .post('/users')
       .send('hello')
       .expect(415);
+    expect(res.body.error.message).to.equal(
+      'Content-type application/x-www-form-urlencoded does not match [application/json].',
+    );
   });
 
   it('returns a user with given id when GET /users/{id} is invoked', async () => {
     const newUser = await userRepo.create(user);
     delete newUser.password;
     delete newUser.orders;
-    // MongoDB returns an id object we need to convert to string
-    // since the REST API returns a string for the id property.
-    newUser.id = newUser.id.toString();
 
     await client.get(`/users/${newUser.id}`).expect(200, newUser.toJSON());
   });
 
-  describe('user product recommendation (service) api', () => {
-    // tslint:disable-next-line:no-any
-    let recommendationService: HttpServer;
-
-    before(async () => {
-      recommendationService = createRecommendationServer();
-      await recommendationService.start();
-    });
+  describe('authentication', () => {
+    it('login returns a valid token', async () => {
+      await client
+        .post('/users')
+        .send(user)
+        .expect(200);
 
-    after(async () => {
-      await recommendationService.stop();
+      const res = await client
+        .post('/users/login')
+        .send({email: user.email, password: user.password})
+        .expect(200);
+
+      const token = res.body.token;
+      expect(token).to.not.be.Null();
+      expect(token).to.be.String();
+      expect(token).to.not.be.empty();
+      const parts = token.split('.');
+      //token has 3 parts separated by '.'
+      expect(parts.length).to.equal(3);
     });
 
-    it('returns product recommendations for a user', async () => {
-      const newUser = await userRepo.create(user);
+    it('login returns an error when invalid email is used', async () => {
       await client
```

Comment:
making http request is slower, please create the user using `const newUser = await userRepo.create(user);` to save time :)

Reply:
@jannyHou 
> making http request is slower, please create the user using const newUser = await userRepo.create(user); to save time :)

Hmm,  I don't agree. The purpose is to test the REST application and its endpoints, not test a repository.
The server needs to start anyways. I don't see the benefit or point of by-passing POST /users.

Reply:
> The purpose is to test the REST application and its endpoints, not test a repository.

Ah true, but this test is for verifying endpoint `Users/login`, therefore it doesn't care about how the user is created, by calling repository API or sending HTTP request, by "making http request is slower" I mean how the user is created. 
Not a big deal though, I will leave it to you to decide how to create the user.

Reply:
I see @jannyHou's point. IIUC, we want to create the user using the less expensive repository operation, then we can use the our http client to make the request to `/users/login` which is what the test aims for. To see this in action, I'd run this test in isolation with both ways and see how much time it took (might have to change the spec to checkmarks instead of dots).

Reply:
@jannyHou , @b-admike explained to me the `specific tests` you wanted me to avoid using `POST /users`. I understand now. I thought you wanted me to avoid `POST /users` for every test.

Reply:
@b-admike Thank you for the explanation! 

> I thought you wanted me to avoid POST /users for every test.

Sorry I meant only those "specific tests", glad you guys figure it out, no misunderstanding now 👏